### PR TITLE
Turese and Lawrence - Participant rank times

### PR DESCRIPTION
**Pull request by lawrencechen14**
- Created view where participants can edit their preferences
- Integrated secret_id so that only people with the id can see the rankings (does not hide right now, but functionality works)
- Created routes for project owner to also be able to change their participants' preferences (can remove later but plausible feature)

### CODE REVIEW

File: features/step_definitions/enter_preferences_steps.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+Given /^a participant to project "(.*)" with email "(.*)" has a secret_id "(.*)"$/ do |project_name, email, secret_id|
```
Comment (kev-chien) [inline features/step_definitions/enter_preferences_steps.rb:1]: I like how you created a new file for this, makes it more organized! Thanks

File: app/controllers/rankings_controller.rb [hunk 2]
```diff
@@ -1,79 +1,123 @@
...
+      end
+    end
+
+    @times.ids.each do |id|
+      rank_num = params[id.to_s].to_i
+      new_ranking = Ranking.new(:rank => rank_num, :participant_id => @participant.id, :project_time_id => id)
+      existing_ranking = Ranking.find_by(:participant_id => @participant.id, :project_time_id => id)
+
+      if existing_ranking
+        Ranking.find_by(:participant_id => @participant.id, :project_time_id => id).update(
```
Comment (kev-chien) [inline app/controllers/rankings_controller.rb]: could you not just do `existing_ranking`.update(` ?

File: app/controllers/rankings_controller.rb [hunk 3]
```diff
@@ -1,79 +1,123 @@
...
-    #def set_ranking
-     # @ranking = Ranking.find(params[:participant_id])
-    #end
+    def set_fields
+      @project = Project.find(params[:project_id])
+      @participant = Participant.find_by(:project_id => params[:project_id], :id => params[:participant_id])
+      @times = @project.project_times
+      @rankings = Ranking.all
+
+      @ranking = Ranking.find_by(:participant_id => params[:participant_id])
```
Comment (kev-chien) [inline app/controllers/rankings_controller.rb:116]: Is this needed? Seems like you only fetch one ranking.

Comment (kev-chien) [reply app/controllers/rankings_controller.rb:116]: Nvm I see that this is referenced in the view.

File: db/schema.rb [hunk 4]
```diff
@@ -21,11 +21,11 @@
   create_table "participants", force: :cascade do |t|
     t.integer  "project_id",     null: false
     t.string   "email",          null: false
-    t.string  "secret_id"
     t.datetime "last_responded"
+    t.string   "secret_id"
   end
 
-  add_index "participants", ["project_id", "email"], name: "index_participants_on_project_id_and_email", unique: true
+  add_index "participants", ["project_id", "email", "last_responded"], name: "index_participants_on_project_id_and_email_and_last_responded", unique: true
```
Comment (kev-chien) [inline db/schema.rb]: this change was undone in Joyce's PR commit. fix merge conflicts carefully!
We don't want to index on all three columns as unique, only need to enforce uniqueness on project_id and email

File: app/views/rankings/show.html.erb [hunk 5]
```diff
@@ -1,5 +1,30 @@
...
+    <tr>
+      <th>Time</th>
+      <th>Rank</th>
+    </tr>
+    </thead>
+
+    <tbody>
+    <% @rankings.each do |ranking| %>
+      <tr>
+<!--        <td> <%#=@times.find(ranking.project_time_id).date_time.to_formatted_s(:long)%></td>-->
```
Comment (kev-chien) [inline app/views/rankings/show.html.erb]: what was the issue here?
each ranking owns a project_time via project_time_id column (in ranking model, there is a `belongs_to project_time`, so have you tried `ranking.project_time.date_time....`?

File: features/enter_preferences.feature [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    And a participant to project "CS169 Sections" with email "empoleon@berkeley.edu" has a secret_id "empoleonsecretid"
+    And a participant to project "CS169 Sections" with email "charizard@berkeley.edu" has a secret_id "charizardsecretid"
+
+
+  Scenario: User enters their preferences for times
+    When I access the time ranking page for project "CS169 Sections" from email "empoleon@berkeley.edu" and secret_id "empoleonsecretid"
+    Then I should see "Please enter your preferences for these times."
+    And I should not see "Access denied"
+    When I choose "Cannot go" for time "Dec 1 2019 10:00 AM"
```
Comment (kev-chien) [inline features/enter_preferences.feature:28]: This is really nice! Thanks for writing the tests and steps to make this possible, it's very flexible and easy to test.

File: app/controllers/rankings_controller.rb [hunk 7]
```diff
@@ -1,79 +1,123 @@
...
   private
     # Use callbacks to share common setup or constraints between actions.
-    #def set_ranking
-     # @ranking = Ranking.find(params[:participant_id])
-    #end
+    def set_fields
+      @project = Project.find(params[:project_id])
+      @participant = Participant.find_by(:project_id => params[:project_id], :id => params[:participant_id])
+      @times = @project.project_times
+      @rankings = Ranking.all
```
Comment (kev-chien) [inline app/controllers/rankings_controller.rb]: should do `@participant.rankings` (participant has_many rankings)
you're fetching all the rankings for the entire database here, just need it for this participant right?

File: app/controllers/rankings_controller.rb [hunk 8]
```diff
@@ -1,79 +1,123 @@
...
+      existing_ranking = Ranking.find_by(:participant_id => @participant.id, :project_time_id => id)
+
+      if existing_ranking
+        Ranking.find_by(:participant_id => @participant.id, :project_time_id => id).update(
+            :rank => rank_num, :participant_id => @participant.id, :project_time_id => id)
+      else
+        new_ranking.save!
+      end
+    end
+
```
Comment (kev-chien) [inline app/controllers/rankings_controller.rb:94]: I think we're missing the feature to update the participant's `last_responded` column when we click submit?

**Review by kev-chien — APPROVED** [review]
Thanks a lot @Turese and @lawrencechen14 ! There's a few minor fixes I've suggested, but overall good work! The form looks really nice! We can leave these fixes for next iteration if desired, but please see to them if you can.

**Review by Aakup — APPROVED** [review]

**Pull request merged**